src/yoroi-wallets/cardano/utxoManager.ts

```diff
@@ -138,3 +63,55 @@ const serializer = (utxo: UtxoModels.Utxo): RawUtxo => ({
 })
 
 export type UtxoManager = Awaited<ReturnType<typeof makeUtxoManager>>
+
+const diffPath = 'diff'
+const safePointPath = 'safe-point'
+
+export const makeUtxoStorage = (storage: Storage) => {
+  const getUtxoDiffToBestBlock = () => storage.getItem(diffPath, parseDiff).then((diff) => diff ?? [])
+  const setUtxoDiffToBestBlock = (utxoDiffToBestBlock: UtxoDiffToBestBlock[]) =>
+    storage.setItem(diffPath, utxoDiffToBestBlock)
+
+  const getUtxoAtSafePoint = () => storage.getItem(safePointPath, parseSafePoint)
+  const setUtxoAtSafePoint = (utxoAtSafePoint: UtxoAtSafePoint) => storage.setItem(safePointPath, utxoAtSafePoint)
+
+  const utxoStorage: UtxoStorage = {
+    getUtxoAtSafePoint,
+    replaceUtxoAtSafePoint: (utxos: Utxo[], lastSafeBlockHash: string) =>
+      setUtxoAtSafePoint({lastSafeBlockHash, utxos}),
+
+    getUtxoDiffToBestBlock,
+    appendUtxoDiffToBestBlock: async (diff: UtxoDiffToBestBlock) => {
+      const currentDiffs = await getUtxoDiffToBestBlock()
+      if (currentDiffs.find((d) => d.lastBestBlockHash === diff.lastBestBlockHash)) return
+      return setUtxoDiffToBestBlock([...currentDiffs, diff])
+    },
+    removeDiffWithBestBlock: async (blockHash: string) => {
+      const currentDiffs = await getUtxoDiffToBestBlock()
+      return setUtxoDiffToBestBlock(currentDiffs.filter((d) => d.lastBestBlockHash !== blockHash))
+    },
+
+    clearUtxoState: () =>
+      Promise.all([storage.removeItem(safePointPath), storage.removeItem(diffPath)]).then(() => undefined),
+  }
+
+  return utxoStorage
+}
+
+const parseSafePoint = (data: unknown) => {
+  const parsed = parseSafe(data)
+  return isSafePoint(parsed) ? parsed : undefined
+}
+const isSafePoint = (data: unknown): data is UtxoAtSafePoint => {
+  const candidate = data as UtxoAtSafePoint
+  return !!candidate && !!candidate.lastSafeBlockHash && Array.isArray(candidate.utxos)
```

Comment:
The type of `lastSafeBlockHash` or the type of items in `utxos` is not checked. This is an issue we face in multiple places of the app. Can we use some parsing library going forward? I've built this https://www.npmjs.com/package/create-typeguard couple of years ago, but I'm sure there are better alternatives out there

Reply:
thats the plan, just havent gotten around to it
https://github.com/colinhacks/zod

Reply:
it wasnt checked before either, i figure its at least a step in the right direction